crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1968,7 +1966,25 @@ impl<'db> TypeInferenceBuilder<'db> {
         expr.map(|expr| self.infer_annotation_expression(expr))
     }
 
+    #[track_caller]
     fn infer_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
+        debug_assert_eq!(
+            self.index.try_expression(expression),
```

Comment:
The `Debug` representations of some of our `Expr` nodes can be huge; this might result in a colossal wall of text being printed to the terminal if the assertion fails. I wonder if just printing the type of the variant might be better here?